### PR TITLE
Generate helpers for `begin_`/`end_` call pairs

### LINKED ISSUE #8: Feature request: Add methods that accept a block and do pushes/pops for you
In a couple projects, I've found myself adding methods like
```crystal
def with_style_color(idx : ImGui::ImGuiCol, col : ImGui::ImVec4, &) : Nil
  ImGui.push_style_color(idx, col)
  yield
  ImGui.pop_style_color
end

def child(str_id : String, size : ImGui::ImVec2 = ImGui::ImVec2.new(0, 0), border : Bool = false, flags : ImGui::ImGuiWindowFlags = ImGui::ImGuiWindowFlags.new(0), &) : Nil
  yield if ImGui.begin_child(str_id, size, border, flags)
  ImGui.end_child
end
```

I think it'd be nice if this shard supported those. I'm happy to add them manually if you like, or if you want to update the generator that'd be awesome. The complication there is that ImGui isn't consistent regarding when `end` methods should be called. For example, you always have to call `end_child` after `begin_child`, but you only call `end_menu` if `begin_menu` returns true. What do you think?

### PR DESCRIPTION
**Pull request by oprypin**
Fixes #8

Thanks @mattrberry for the idea and collaboration!

### CODE REVIEW

**Comment by oprypin** [comment]
I haven't finished the part about converting the demos. But all the main code generation is finished. See *obj.cr*.
Feedback about naming is still accepted for sure :)

**Comment by oprypin** [comment]
Ah and also something I briefly thought about: should the `pop`s perhaps be in an `ensure` block? But probably not, it's likely to break ImGui even more, if something is already broken.

**Comment by oprypin** [comment]
And: should the return value be propagated from the `yield`? Currently these have no return value.

File: generate.cr [hunk 1]
```diff
@@ -467,13 +467,39 @@ class COverload
...
       yield call
       yield (assert to_tuple(outp2)) unless outp2.empty? || outp2 == ["result"]
       yield %(  end)
       yield %(  {% end %}) if as_datatype
+
+      return if inside_class
+      antonyms = {"push" => "pop", "begin" => "end"}
+      names = {"push" => "with", "begin" => "within"}
+      always_pop_names = %w[begin begin_child begin_child_frame begin_disabled begin_group begin_tooltip push_allow_keyboard_focus push_button_repeat push_clip_rect push_clip_rect_full_screen push_font push_id push_item_width push_style_color push_style_var push_texture_id push_text_wrap_pos]
+      cond_pop_names = %w[begin_combo begin_drag_drop_source begin_drag_drop_target begin_list_box begin_main_menu_bar begin_menu begin_menu_bar begin_popup begin_popup_context_item begin_popup_context_void begin_popup_context_window begin_popup_modal begin_tab_bar begin_tab_item begin_table tree_node]
```
Comment (mattrbeck) [inline generate.cr]: Do we also want to add `tree_push`? There aren't any uses of it in the demo, but I think it should be possible to use tree_push/tree_pop

**Comment by mattrbeck** [comment]
> Ah and also something I briefly thought about: should the `pop`s perhaps be in an `ensure` block? But probably not, it's likely to break ImGui even more, if something is already broken.

Hmm that's just a wonky state overall. I think either option is fine as long as it's documented. I may lean slightly towards using an `ensure` block because my understanding of these blocks is that they'll remove any place the user needs to push/pop or begin/end. However, I'd still re-raise the exception because it's still up to the user to handle the error beyond that point. Just my initial thought

**Comment by mattrbeck** [comment]
> And: should the return value be propagated from the `yield`? Currently these have no return value.

I don't see a huge value in this since the user knows the condition is true whenever the block is entered. That said, I don't see any harm in yielding the value, so maybe it'd be good to do it just in case it's helpful in the way the user wants to write their code?

File: generate.cr [hunk 2]
```diff
@@ -467,13 +467,39 @@ class COverload
...
+
+      if (cond_pop = cond_pop_names.includes?(def_name)) || always_pop_names.includes?(def_name)
+        cond_call = !cond_pop && (ret_s == "Bool")
+        if def_name == "tree_node"
```
Comment (mattrbeck) [inline generate.cr]: I just scrolled through imgui.h quickly and it seems like `TreePush` and `TreePop` are the only ones that don't have the push/pop at the beginning of the method name, so hopefully `TreePush` and `TreeNode` are the only two special cases needed here :+1:

File: generate.cr (hunk 1 above)
Comment (oprypin) [reply generate.cr]: Yes I'll add that. Just not sure how to name that method 🤔
`with_tree` is what I was thinking mainly

File: src/obj.cr [hunk 3]
```diff
@@ -197,6 +197,12 @@ module ImGui
     LibImGui.Begin(name, p_open, flags)
   end
 
+  # Calls `begin`, conditionally yields to the block, then calls `end`.
+  def self.within(name : String, p_open : Bool* = Pointer(Bool).null, flags : ImGuiWindowFlags = ImGuiWindowFlags.new(0)) : Nil
```
Comment (oprypin) [inline src/obj.cr]: I am strongly considering removing the prefix from methods such as `within_menu_bar` &rarr; `menu_bar`.
There's just the problem that `within` would be renamed to nothing according to that scheme 😅
What does `begin` mean - begin a window? Should we name this `window` then? Or what do you think?

**Comment by mattrbeck** [comment]
> I am strongly considering removing the prefix from methods such as `within_menu_bar` → `menu_bar`. There's just the problem that `within` would be renamed to nothing according to that scheme sweat_smile What does `begin` mean - begin a window? Should we name this `window` then? Or what do you think?

Assuming that's the decision, I agree with `begin` → `window` since that's how it's described in imgui. My only thought about dropping the `within`, though, is that the name may not be unique. This holds true for the `combo` case, which is now another method overload. I think I actually still prefer it with the `within` dropped, but just something to think about.

If we're dropping the `within`, what do you think about the `with` on the push/pop blocks? I think I favor having it since it's just a little more explicit, but I'm curious what you think

**Comment by oprypin** [comment]
Yes. Well, it's all finished now. Please give it a try, maybe check out how the new functions and the demo code look.

**Comment by mattrbeck** [comment]
Just gave it a try with my own code. I like it! :shipit:

**Pull request merged**